network/src/peers_registry.rs

```diff
@@ -112,235 +113,283 @@ impl PeerConnection {
+        group.push(peer_id);
+        if group.len() > largest_group_len {
+            largest_group_len = group.len();
+            largest_group = group_name;
+        }
+    }
+    groups[&largest_group].clone()
 }
 
 impl PeersRegistry {
     pub fn new(
         peer_store: Arc<RwLock<Box<PeerStore>>>,
-        max_incoming: u32,
-        max_outgoing: u32,
+        max_inbound: u32,
+        max_outbound: u32,
         reserved_only: bool,
+        reserved_peers: Vec<PeerId>,
     ) -> Self {
-        let deny_list = PeersDenyList::new();
+        let mut reserved_peers_set =
+            FnvHashSet::with_capacity_and_hasher(reserved_peers.len(), Default::default());
+        for reserved_peer in reserved_peers {
+            reserved_peers_set.insert(reserved_peer);
+        }
         PeersRegistry {
             peer_store,
-            peer_connections: Default::default(),
-            max_incoming,
-            max_outgoing,
+            peers: Default::default(),
+            reserved_peers: reserved_peers_set,
+            max_inbound,
+            max_outbound,
             reserved_only,
-            deny_list,
         }
     }
 
     #[allow(clippy::needless_lifetimes)]
     #[inline]
     pub fn get_peer_id<'a>(&'a self, peer_index: PeerIndex) -> Option<&'a PeerId> {
-        self.peer_connections.get_peer_id(peer_index)
+        self.peers.get_peer_id(peer_index)
     }
 
-    // registry a new peer
-    #[allow(clippy::needless_pass_by_value)]
-    pub fn new_peer(&mut self, peer_id: PeerId, endpoint: Endpoint) -> Result<(), Error> {
-        if self.peer_connections.get(&peer_id).is_some() {
+    pub fn is_reserved(&self, peer_id: &PeerId) -> bool {
+        self.reserved_peers.contains(&peer_id)
+    }
+
+    pub fn accept_inbound_peer(&mut self, peer_id: PeerId, addr: Multiaddr) -> Result<(), Error> {
+        if self.peers.get(&peer_id).is_some() {
             return Ok(());
         }
-        let is_reserved = self.peer_store.read().is_reserved(&peer_id);
+        if !self.is_reserved(&peer_id) {
+            if self.reserved_only {
+                return Err(ErrorKind::InvalidNewPeer(format!(
+                    "We are in reserved_only mode, rejected non-reserved peer {:?}",
+                    peer_id
+                ))
+                .into());
+            }
+            let candidate_score = {
+                let peer_store = self.peer_store.read();
+                if peer_store.is_banned(&peer_id) {
+                    return Err(
+                        ErrorKind::InvalidNewPeer(format!("peer {:?} is denied", peer_id)).into(),
+                    );
+                }
+                peer_store.peer_score_or_default(&peer_id)
+            };
+            let connection_status = self.connection_status();
+            // check peers connection limitation
+            if connection_status.unreserved_inbound >= self.max_inbound
```

Comment:
`connection_status` function is only used here, I think it's better to extract to a new function `over_max_inbound_limit` and remove `connection_status`

Reply:
We have several unimplemented eviction strategies which described in RFC-0007, this function would be helpful when applying that logic.